obspy/core/tests/test_resource_identifier.py

```diff
@@ -0,0 +1,611 @@
+        """
+        # This one contains all valid characters. It should pass the
+        # validation.
+        res_id = (
+            "smi:abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+            "1234567890-.*()_~'/abcdefghijklmnopqrstuvwxyzABCDEFGHIKLMNOPQR"
+            "STUVWXYZ0123456789-.*()_~'+?=,;&")
+        res = ResourceIdentifier(res_id)
+        self.assertEqual(res_id, res.get_quakeml_uri())
+        # The id has to valid from start to end. Due to the spaces this cannot
+        # automatically be converted to a correct one.
+        res_id = "something_before smi:local/something  something_after"
+        res = ResourceIdentifier(res_id)
+        self.assertRaises(ValueError, res.get_quakeml_uri)
+        # A colon is an invalid character.
+        res_id = "smi:local/hello:yea"
+        res = ResourceIdentifier(res_id)
+        self.assertRaises(ValueError, res.get_quakeml_uri)
+        # Space as well
+        res_id = "smi:local/hello yea"
+        res = ResourceIdentifier(res_id)
+        self.assertRaises(ValueError, res.get_quakeml_uri)
+        # Dots are fine
+        res_id = "smi:local/hello....yea"
+        res = ResourceIdentifier(res_id)
+        self.assertEqual(res_id, res.get_quakeml_uri())
+        # Hats not
+        res_id = "smi:local/hello^^yea"
+        res = ResourceIdentifier(res_id)
+        self.assertRaises(ValueError, res.get_quakeml_uri)
+
+    def test_resource_id_valid_quakemluri(self):
+        """
+        Test that a resource identifier per default (i.e. no arguments to
+        __init__()) gets set up with a QUAKEML conform ID.
+        """
+        rid = ResourceIdentifier()
+        self.assertEqual(rid.id, rid.get_quakeml_uri())
+
+    def test_de_referencing_when_object_goes_out_of_scope(self):
+        """
+        Tests that objects that have no more referrer are no longer stored in
+        the reference dictionary.
+        """
+        r_dict = self.id_order
+        t1 = UTCDateTime(2010, 1, 1)  # test object
+        rid = 'a'  # test resource id
+
+        # Create object and assert the reference has been created.
+        r1 = ResourceIdentifier(rid, referred_object=t1)
+        self.assertEqual(r1.get_referred_object(), t1)
+        self.assertTrue(r1._resource_key in r_dict)
+        # Deleting the object should remove the reference.
+        r_dict_len = len(r_dict)
+        del r1
+        self.assertEqual(len(r_dict), r_dict_len - 1)
+        # Now create two equal references.
+        r1 = ResourceIdentifier(rid, referred_object=t1)
+        r2 = ResourceIdentifier(rid, referred_object=t1)
+        self.assertEqual(r1.get_referred_object(), t1)
+        # Deleting one should not remove the reference.
+        del r1
+        self.assertEqual(r2.get_referred_object(), t1)
+        self.assertIn(r2._resource_key, r_dict)
+        # Deleting the second one should (r_dict should not be empty)
```

Comment:
s/not/now

Reply:
Sorry, I am not sure what you mean.

Reply:
Sorry I should have been more clear. `s/SEARCH/REPLACE` is `ack` syntax to search and replace strings. It should be `r_dict should now be empty`.